### PR TITLE
Improve docs on scope and target platforms.

### LINKED ISSUE #57: Make scope more explicit in the charter
The charter currently says:

> OmnICU will provide an ECMA-402-compatible API surface in the target client-side platforms

and:

> **What if clients need a feature that is not in ECMA-402?**
> Clients of OmnICU may need features beyond those recommended by ECMA-402. The subcommittee is not ruling out the option of adding additional features in the same style as ECMA-402 to cover additional client needs. The details for how to determine what features belong in OmnICU that aren't already in ECMA-402 will be discussed at a future time.

The caption in ecosystem.md says:

> This document tracks the crates that already exist in the ecosystem that cover functionality that we may wish to cover in OmnICU.

I added the word "may" in #41.

I think it's important that we be more explicit about the use cases that ICU4X is supporting.  This will guide our discussions, such as https://github.com/unicode-org/icu4x/pull/43#discussion_r413329507, when deciding whether a certain API or functional unit belongs in ICU4X.

We could start by making an explicit list of use cases that warrant APIs and functional units not covered by 402, and adding that list to the charter.  It might be best to do this on a case-by-case basis: if proposing a feature not explicitly sanctioned by the charter, then propose a change to the charter adding the corresponding use case to the charter, such that we can agree on that change in the subcommittee meeting.

@zbraniecki

### PR DESCRIPTION
**Pull request by zbraniecki**
Fixes #57.

### CODE REVIEW

File: docs/charter.md [hunk 1]
```diff
@@ -1,7 +1,11 @@
 ICU4X Charter
 ==============
 
-ICU4X is a new project whose objective is to solve the needs of clients who wish to provide client-side i18n for their products in resource-constrained environments.
+ICU4X is a project whose objective is to solve the needs of clients who wish to provide client-side i18n for their products in resource-constrained environments.
```
Comment (srl295) [inline docs/charter.md:4]: "objective to solve the needs…" is perhaps too many levels of indirection. How about something like, "The ICU4X project provides client-side i18n for resource-constrainied environments"?

File: docs/charter.md [hunk 2]
```diff
@@ -1,7 +1,11 @@
...
+
+ICU4X derives from the experience of ICU4C/ICU4J and ECMA402 design decisions.
```
Comment (srl295) [inline docs/charter.md:6]: ICU4X builds on the design and API decisions of ICU4C/ICU4X and ECMA-402?

(should use consistent naming for ECMA-402 …)

Also maybe endnote links to ICU[CJ], ECMA, etc…

File: docs/charter.md [hunk 3]
```diff
@@ -10,7 +14,43 @@ ICU4X will be built from the start with several key design constraints:
 3. Availability and ease of use in multiple programming languages.
 4. Written by i18n experts to encourage best practices.
 
-ICU4X will provide an API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+# Scope
+
+ICU4X aims to provide a set of APIs sufficient to internationalize basic modern software.
```
Comment (srl295) [inline docs/charter.md:19]: "aim" and "objective" seem somewhat speculative, I know this is still early, but shouldn't this be written with the 'final' product in mind?

File: docs/charter.md [hunk 4]
```diff
@@ -10,7 +14,43 @@ ICU4X will be built from the start with several key design constraints:
...
+
+With the focus on client-side, it will initially focus on the scope covered by the ECMA-402 extended to cover needs of more low level systems such as Gecko or Fuchsia.
```
Comment (srl295) [inline docs/charter.md:21]:
```suggestion
With the client-side focus, the initial scope is that covered by ECMA-402, with some extensions to cover needs of low level systems such as Gecko or Fuchsia.
```

File: docs/charter.md [hunk 5]
```diff
@@ -10,7 +14,43 @@ ICU4X will be built from the start with several key design constraints:
...
+
+The current target is to cover all of ECMA-402 needs, and a modern subset of ICU4X APIs needed for software platforms such as Fuschia and Gecko.
+
+One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.
+
+For ICU4C APIs and other internationalization needs that will not end up being included, ICU4X aims to provide a sufficient access to building blocks to allow third-party libraries to be developed that can interoperate with ICU4X with minimum data and logic duplication.
+
+In some rare cases, ICU4X may diverge from both ECMA-402, and ICU4C. Such diversions should be the result of improved API design and should allow users to cover the same needs with the upgraded APIs.
```
Comment (srl295) [inline docs/charter.md:29]: first paragraph duplicates previous one

```suggestion
Two of the benchmarks for inclusion in ICU4X are: (1) if the API requires Unicode/CLDR data and (2) if it would benefit from being managed with other APIs that rely on the same version of Unicode and CLDR data.

For such APIs and functionality that will not end up being included, ICU4X aims to provide the building blocks to enable third-party libraries to be developed which depend on and interoperate with ICU4X, with minimum data and logic duplication.

In some rare cases, ICU4X may diverge from both ECMA-402, and ICU4C/ICU4J. Such diversions should be the result of improved API design and should allow users to cover the same needs with the upgraded APIs.
```

File: docs/charter.md [hunk 6]
```diff
@@ -10,7 +14,43 @@ ICU4X will be built from the start with several key design constraints:
...
+* Rust
+* JavaScript
+* Objective-C
+* Java
+* Dart
+* C++
+
+Referential platforms and languages may change over time as the industry developers and should represent the current needs of the industry.
+
+A viable subset of ICU4X will be targeting the `no_std` support and in the future we may explore the `no_std+alloc` compatibility for viable components.
```
Comment (srl295) [inline docs/charter.md:53]: (Maybe links to `no_std` / `no_std+alloc` documentation ? )

File: docs/charter.md (hunk 4 above)
Comment (sffc) [reply docs/charter.md:21]: My version:

> The scope of ICU4X is the set of functionality covered by ECMA-402, with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of ECMA-402.

File: docs/charter.md [hunk 7]
```diff
@@ -10,7 +14,43 @@ ICU4X will be built from the start with several key design constraints:
...
-ICU4X will provide an API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+# Scope
+
+ICU4X aims to provide a set of APIs sufficient to internationalize basic modern software.
+
+With the focus on client-side, it will initially focus on the scope covered by the ECMA-402 extended to cover needs of more low level systems such as Gecko or Fuchsia.
+
+The current target is to cover all of ECMA-402 needs, and a modern subset of ICU4X APIs needed for software platforms such as Fuschia and Gecko.
+
+One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.
```
Comment (sffc) [inline docs/charter.md:25]: This is the most important paragraph in my opinion.  We should list out what all of the benchmarks are.  Suggestions:

- Useful to a broad audience of client-side applications
- Depends on Unicode or CLDR data or algorithms
- Benefits from being versioned alongside Unicode and CLDR data

Comment (srl295) [reply docs/charter.md:25]: i have a suggested change below 👇

**Comment by zbraniecki** [comment]
Thank you @sffc and @srl295 ! I applied your feedback mixing it in a few places for greater effect.

Please, re-review.

File: docs/charter.md [hunk 8]
```diff
@@ -1,28 +1,70 @@
 ICU4X Charter
 ==============
 
-ICU4X is a new project whose objective is to solve the needs of clients who wish to provide client-side i18n for their products in resource-constrained environments.
+The ICU4X project designs and develops a set of modular internationalization components suitable for use on the client-side and in resource-constrained environments.
```
Comment (sffc) [inline docs/charter.md:4]: "client-side", with a hyphen, is an adjective.  You should say either "for use on the client side and in resource-constrained environments" or "for use in client-side or resource-constrained environments".

(I'm being extra picky since this is the first sentence of the charter)

File: docs/charter.md [hunk 9]
```diff
@@ -1,28 +1,70 @@
...
 
-ICU4X will be built from the start with several key design constraints:
+ICU4X builds on the design and API decisions of [`ICU4C`]/[`ICU4J`] and [`ECMA-402`].
```
Comment (sffc) [inline docs/charter.md:6]: Optional: Using backticks seems like an odd style choice.  I would just write `[ICU4C]` (and ICU4J, ECMA-402).  They will turn to blue links, which I think is enough to highlight them from the background.

File: docs/charter.md [hunk 10]
```diff
@@ -1,28 +1,70 @@
...
 2. Pluggable locale data.
 3. Availability and ease of use in multiple programming languages.
 4. Written by i18n experts to encourage best practices.
 
-ICU4X will provide an API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+## Scope
+
+The scope of ICU4X is the set of functionality covered by [`ECMA-402`], with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of [`ECMA-402`].
+
+One of the benchmarks on decisions related to whether a given API belongs to ICU4X will be if it requires Unicode/CLDR data and would benefit from being managed together with other APIs that rely on the same version of Unicode and CLDR data.
```
Comment (sffc) [inline docs/charter.md:21]: Suggestion: This paragraph seems redundant with the third bullet point.  I would probably remove it.

File: docs/charter.md [hunk 11]
```diff
@@ -1,28 +1,70 @@
...
+* Rust
+* JavaScript
+* Objective-C
+* Java
+* Dart
+* C++
+
+Referential platforms and languages may change over time as the industry developers and should represent the current needs of the industry.
+
+A viable subset of ICU4X will be targeting the [`no_std`] support and in the future we may explore the [`no_std`]+[`alloc`] compatibility for viable components.
```
Comment (sffc) [inline docs/charter.md:55]: Suggestion/Optional: Link to https://github.com/unicode-org/icu4x/issues/77

File: docs/charter.md [hunk 12]
```diff
@@ -52,7 +94,7 @@ A new project allows us to build in the client-side needs from day 1, including
 
 ### How will you actually implement ICU4X?
 
-Internally, the first draft of ICU4X will be built in a [no_std Rust Environment](https://rust-embedded.github.io/book/intro/no-std.html), with plans to use either FFI or [WebAssembly](https://webassembly.org/) for porting the code to the other target platforms; however, we are leaving the door open for other options such as transpilation that can produce code directly.  For example, we expect Rust clang/llvm tools to stabilize in the future, which could be the basis of a potential transpilation effort.  We are also looking at a Lisp-like source that can generate high-level code.
+Internally, the first draft of ICU4X will be built in a [`no_std`], with plans to use either FFI or [WebAssembly](https://webassembly.org/) for porting the code to the other target platforms; however, we are leaving the door open for other options such as transpilation that can produce code directly.  For example, we expect Rust clang/llvm tools to stabilize in the future, which could be the basis of a potential transpilation effort.  We are also looking at a Lisp-like source that can generate high-level code.
```
Comment (sffc) [inline docs/charter.md:97]: `"in a [no_std] Rust environment"` (you lost the "Rust environment")

**Comment by zbraniecki** [comment]
Ready for re-review

**Comment by zbraniecki** [comment]
@srl295 @nciric - can you please either review or remove yourself from reviewers?

File: docs/charter.md [hunk 13]
```diff
@@ -1,28 +1,68 @@
...
-ICU4X will provide an API surface in the target client-side platforms, including the web platform, iOS, Android, WearOS, WatchOS, Flutter, and Fuchsia, supported in programming languages including Rust, JavaScript, Objective-C, Java, Dart, and C++.
+## Scope
+
+The scope of ICU4X is the set of functionality covered by [ECMA-402], with extensions to cover the needs of clients including Gecko and Fuchsia that are consistent with the spirit of [ECMA-402].
+
+APIs included in ICU4X will:
+
+* Be useful to a broad audience of client-side applications
+* Depend on Unicode or CLDR data or algorithms
+* Benefits from being versioned alongside Unicode and CLDR data
```
Comment (nciric) [inline docs/charter.md:25]: Benefit vs Benefits

File: docs/charter.md [hunk 14]
```diff
@@ -38,7 +78,7 @@ ICU4X will have an independent code base from ICU, and will operate independentl
 
 No!
```
Comment (nciric) [inline docs/charter.md:79]: No! what?

Comment (sffc) [reply docs/charter.md]: No to answer the question right above it.

File: docs/charter.md [hunk 15]
```diff
@@ -36,9 +76,7 @@ ICU4X will have an independent code base from ICU, and will operate independentl
 
 ### Is ICU4X going to replace ICU?
 
-No!
```
Comment (sffc) [inline docs/charter.md:39]: Why did you delete this line? Please add it back unless you think people need to read the full paragraph in order to answer this question.

Comment (zbraniecki) [reply docs/charter.md:39]: Because it doesn't add a lot of value and doesn't feel like an official document.
A paragraph is not long so I think a more neuanced answer should be ok. 

I can bring it back, if you are strongly in favor of keeping it. Just sync with @nciric on how it should look like.

Comment (sffc) [reply docs/charter.md:39]: OK, if @zbraniecki and @nciric think the line is not necessary, let's remove it.  I just wanted to make sure it wasn't removed accidentally.
